tests/test_coordinated_workers/test_coordinator.py

```diff
@@ -281,3 +284,103 @@ def test_tracing_receivers_urls(
+        }.items()
+    }
+    requires_relations["my-s3"] = testing.Relation(
+        "my-s3",
+        interface="s3",
+        remote_app_data={
+            "endpoint": "s3",
+            "bucket": "foo-bucket",
+            "access-key": "my-access-key",
+            "secret-key": "my-secret-key",
+        },
+    )
+    requires_relations["cluster_worker0"] = testing.Relation(
+        "my-cluster",
+        remote_app_name="worker0",
+        remote_app_data=ClusterRequirerAppData(role="read").dump(),
+    )
+    requires_relations["cluster_worker1"] = testing.Relation(
+        "my-cluster",
+        remote_app_name="worker1",
+        remote_app_data=ClusterRequirerAppData(role="write").dump(),
+    )
+    requires_relations["cluster_worker2"] = testing.Relation(
+        "my-cluster",
+        remote_app_name="worker2",
+        # remote_app_data=ClusterRequirerAppData(role="backend").dump(),
+    )
+
+    provides_relations = {
+        endpoint: testing.Relation(endpoint=endpoint, interface=interface["interface"])
+        for endpoint, interface in {
+            "my-dashboards": {"interface": "grafana_dashboard"},
+            "my-metrics": {"interface": "prometheus_scrape"},
+        }.items()
+    }
+
+    return testing.State(
+        containers={
+            testing.Container("nginx", can_connect=True),
+            testing.Container("nginx-prometheus-exporter", can_connect=True),
+        },
+        relations=list(requires_relations.values()) + list(provides_relations.values()),
+    )
+
+
+@pytest.fixture()
+def cluster_charm(request):
+    class MyCluster(ops.CharmBase):
+        META = {
+            "name": "foo-app",
+            "requires": {
+                "my-certificates": {"interface": "certificates"},
+                "my-cluster": {"interface": "cluster"},
+                "my-logging": {"interface": "loki_push_api"},
+                "my-charm-tracing": {"interface": "tracing", "limit": 1},
+                "my-workload-tracing": {"interface": "tracing", "limit": 1},
+                "my-s3": {"interface": "s3"},
+            },
+            "provides": {
+                "my-dashboards": {"interface": "grafana_dashboard"},
+                "my-metrics": {"interface": "prometheus_scrape"},
+            },
+            "containers": {
+                "nginx": {"type": "oci-image"},
+                "nginx-prometheus-exporter": {"type": "oci-image"},
+            },
+        }
+
+        def __init__(self, framework: ops.Framework):
+            super().__init__(framework)
+            # Note: Here it is a good idea not to use context mgr because it is "ops aware"
+            self.cluster = ClusterProvider(
+                charm=self,
+                # Roles were take from loki-coordinator-k8s-operator
+                roles=frozenset({"all", "read", "write", "backend"}),
+                endpoint="my-logging",
+            )
+
+    return MyCluster
+
+
+def test_invalid_databag_content(
```

Comment:
as it is this test is rather brittle: in general it's not a great practice to test logging output as that is subject to changing.
Is there something else we can assert to verify the right behaviour?
IIRC the charm will set a Waiting/Blocked status if the remote databag is invalid. That is more robust, and does not require you to use the context manager and poke into the charm's internals (which are also brittle)

Reply:
I have switch to check the unit status, but using MyCoordinator CluserProvider was a bit complexe and calling the function directly still seems the best. I am opened to look further, but would require a bit of your help to understand how to acheive it.